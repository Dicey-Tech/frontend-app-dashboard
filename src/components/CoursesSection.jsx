import React, { useState, useEffect } from 'react';
import { Container, Row } from '@edx/paragon';
import { faSearch } from '@fortawesome/free-solid-svg-icons';
import { getConfig } from '@edx/frontend-platform';
import SectionTitle from './SectionTitle';
import ExploreCard from './cards/ExploreCard';
import LmsApiService from '../app/services/LmsApiService';
import DiceySpinner from './DiceySpinner';
import DashboardCard from './cards/DashboardCard';

const fetchStudentCourses = async () => {
  const response = await LmsApiService.fetchStudentCourses();
  const { data } = response;
  const result = [];
  const courseEnrollments = data.map((element) => ({
    course_id: element.course_details.course_id,
  }));

  /* now fetch the individual course information including the image */
  /* eslint-disable no-restricted-syntax */
  /* eslint-disable no-await-in-loop */
  for (const element of courseEnrollments) {
    try {
      const course = await LmsApiService.fetchCourseInfo(element.course_id);

      result.push({
        course_id: element.course_id,
        name: course.data.name,
        description: course.data.short_description,
        media: course.data.media.image.small,
        start: Date.parse(course.data.start),
      });
    } catch {
      /* Skip that course */
    }
  }
  /* eslint-enable no-await-in-loop */
  /* eslint-enable no-restricted-syntax */
  return result;
};

const CoursesSection = () => {
  const [courses, setCourses] = useState([]);
  const [isLoading, setIsLoading] = useState(false);

  useEffect(() => {
    setIsLoading(true);
    fetchStudentCourses().then((data) => {
      setCourses(data.sort((elemA, elemB) => elemB.start - elemA.start));
    })
      .catch((e) => {
        throw e;
      })
      .finally(() => {
        setIsLoading(false);
      });
  }, []);

  /* eslint-disable no-nested-ternary */
  const courseCards = isLoading ? <DiceySpinner />
    : courses.length === 0 ? (
      <div key="no-classrooms" className="col align-self-center d-inline">
        <h3>You are not enrolled in any courses</h3>
      </div>
    )
      : (
        <div key="course-list" className="card-scroll-region col">
          {courses.map((element) => {
            const lmsRoot = getConfig().LMS_BASE_URL;
            return (
              <DashboardCard
                key={element.course_id}
                name={element.name}
                description={element.description}
                media={element.media}
                url={`${lmsRoot}/courses/${element.course_id}/course`}
                actionBtnText="Start Now"
              />
            );
          })}
        </div>
      );
  /* eslint-enable no-nested-ternary */

  return (
    <Container className="section-container">
      <SectionTitle>My Courses</SectionTitle>
      <Row className="card-row px-2">
        {(getConfig().EXPLORE_COURSES_URL || '') !== ''
          ? <ExploreCard text="Find New Courses" icon={faSearch} /> : null}
        {courseCards}
      </Row>
    </Container>
  );
};

export default CoursesSection;
