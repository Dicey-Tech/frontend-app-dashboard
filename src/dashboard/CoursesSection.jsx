import React from 'react';
import PropTypes from 'prop-types';

import {
  Container, DataTable, CardView, TableFooter,
} from '@edx/paragon';
import { getConfig } from '@edx/frontend-platform';

import { SectionTitle, DashboardCard, DiceySpinner } from '../common-components';

const INITIAL_STATE = {
  pageSize: 6,
  pageIndex: 0,
};

const COLUMNS = [
  {
    Header: 'Title',
    accessor: 'name',
  },
  {
    Header: 'Description',
    accessor: 'description',
  },
];

const COLUMN_SIZES = { xs: 12, lg: 6, xl: 2 };

const CoursesSection = ({ courses, isReady, hasEnrollments }) => {
  /* eslint-disable no-nested-ternary */
  const courseCards = !isReady ? <DiceySpinner />
    : !hasEnrollments ? (
      <div key="no-classrooms" className="col align-self-center d-inline">
        <h3>You are not enrolled in any courses</h3>
        <p>Getting started with {getConfig().SITE_NAME} is easy. Simply find a project from the catalog</p>
      </div>
    )
      : (
        <DataTable
          isPaginated
          isSortable
          data={courses}
          itemCount={courses.length}
          initialState={INITIAL_STATE}
          columns={COLUMNS}
        >
          <CardView CardComponent={DashboardCard} columnSizes={COLUMN_SIZES} />
          <TableFooter />
        </DataTable>
      );

  return (
    <>
      <Container className="section-container">
        <SectionTitle>My Projects</SectionTitle>
        {courseCards}
      </Container>
    </>
  );
};

CoursesSection.propTypes = {
  isReady: PropTypes.bool,
  hasEnrollments: PropTypes.bool,
  courses: PropTypes.arrayOf(PropTypes.object),
};

CoursesSection.defaultProps = {
  isReady: false,
  hasEnrollments: false,
  courses: [],
};

export default CoursesSection;
