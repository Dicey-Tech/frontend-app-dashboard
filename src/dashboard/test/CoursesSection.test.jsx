import React from 'react';
import renderer from 'react-test-renderer';

import CoursesSection from '../CoursesSection';

describe('<CoursesSection />', () => {
  let props = {};

  beforeEach(() => {
    props = {
      isReady: true,
      hasEnrollments: true,
      courses: [
        {
          name: 'Project X',
          short_description: 'Description of Project X',
          course_id: 'course-v1:DiceyTech+PRJX001+PROJECT_X',
          media: {
            image: {
              small: 'images_course_image.jpg',
            },
          },
        },
      ],
    };
  });

  it('should render course section if a list of courses is provided', () => {
    const tree = renderer.create(
      <CoursesSection {...props} />,
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('should render a message if the user is not enrolled in any course', () => {
    props = {
      isReady: true,
      hasEnrollments: false,
      courses: [],
    };

    const tree = renderer.create(
      <CoursesSection {...props} />,
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
