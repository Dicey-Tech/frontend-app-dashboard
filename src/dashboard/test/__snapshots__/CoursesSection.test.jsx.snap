// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CoursesSection /> should render a message if the user is not enrolled in any course 1`] = `
<div
  className="section-container container-fluid"
>
  <div
    className="align-items-center section-title row"
  >
    <div
      className="col"
    >
      <h2>
        My Projects
      </h2>
    </div>
  </div>
  <div
    className="col align-self-center d-inline"
  >
    <h3>
      You are not enrolled in any courses
    </h3>
    <p>
      Getting started with 
      edX
       is easy. Simply find a project from the catalog
    </p>
  </div>
</div>
`;

exports[`<CoursesSection /> should render course section if a list of courses is provided 1`] = `
<div
  className="section-container container-fluid"
>
  <div
    className="align-items-center section-title row"
  >
    <div
      className="col"
    >
      <h2>
        My Projects
      </h2>
    </div>
  </div>
  <div
    className="pgn__data-table-layout-wrapper"
  >
    <div
      className="pgn__data-table-layout-main"
    >
      <div
        className="pgn__data-table-wrapper"
      >
        <div
          className="pgn__card-grid pgn__data-table-card-view"
        >
          <div
            className="row"
          >
            <div
              className="col-xl-2 col-lg-6 col-12"
            >
              <div
                className="pgn__card clickable card"
                tabIndex="0"
              >
                <a
                  href="http://localhost:18000/courses/course-v1:DiceyTech+PRJX001+PROJECT_X/course"
                >
                  <div
                    className="card-image pgn__card-wrapper-image-cap vertical"
                  >
                    <img
                      alt="Project Image"
                      className="pgn__card-image-cap"
                      src="images_course_image.jpg"
                    />
                  </div>
                  <div
                    className="pgn__card-header"
                  >
                    <div
                      className="pgn__card-header-content"
                    >
                      <div
                        className="pgn__card-header-title-md"
                      >
                        Project X
                      </div>
                    </div>
                  </div>
                  <div
                    className="pgn__card-section text-wrap muted-link"
                  >
                    <p>
                      Description of Project X
                    </p>
                  </div>
                </a>
              </div>
            </div>
          </div>
        </div>
        <div
          className="pgn__data-table-footer"
        >
          <div>
            Showing 
            1
             of 
            1
            .
          </div>
          <nav
            aria-label="table pagination"
            className="pagination-minimal"
          >
            <div
              aria-atomic={true}
              aria-live="polite"
              aria-relevant="text"
              className="sr-only"
            >
              Page 1, Current Page, of 1
            </div>
            <ul
              className="pagination"
            >
              <li
                className="page-item disabled"
              >
                <button
                  aria-label="Previous"
                  className="btn-icon btn-icon-primary btn-icon-md previous page-link"
                  disabled={true}
                  onClick={[Function]}
                  tabIndex="-1"
                  type="button"
                >
                  <span
                    className="btn-icon__icon-container"
                  >
                    <span
                      className="pgn__icon btn-icon__icon"
                    >
                      <svg
                        aria-hidden={true}
                        fill="none"
                        focusable={false}
                        height={24}
                        role="img"
                        viewBox="0 0 24 24"
                        width={24}
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="M17.51 3.87L15.73 2.1 5.84 12l9.9 9.9 1.77-1.77L9.38 12l8.13-8.13z"
                          fill="currentColor"
                        />
                      </svg>
                    </span>
                  </span>
                </button>
              </li>
              <li
                className="page-item disabled"
              >
                <button
                  aria-label="Next"
                  className="btn-icon btn-icon-primary btn-icon-md next page-link"
                  disabled={true}
                  onClick={[Function]}
                  tabIndex="-1"
                  type="button"
                >
                  <span
                    className="btn-icon__icon-container"
                  >
                    <span
                      className="pgn__icon btn-icon__icon"
                    >
                      <svg
                        aria-hidden={true}
                        fill="none"
                        focusable={false}
                        height={24}
                        role="img"
                        viewBox="0 0 24 24"
                        width={24}
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="M6.49 20.13l1.77 1.77 9.9-9.9-9.9-9.9-1.77 1.77L14.62 12l-8.13 8.13z"
                          fill="currentColor"
                        />
                      </svg>
                    </span>
                  </span>
                </button>
              </li>
            </ul>
          </nav>
        </div>
      </div>
    </div>
  </div>
</div>
`;
